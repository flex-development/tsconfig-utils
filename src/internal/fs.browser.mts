/**
 * @file Internal - fs/browser
 * @module tsconfig-utils/internal/fs/browser
 */

import type { FileSystem } from '@flex-development/tsconfig-utils'

/**
 * File system API.
 *
 * @const {FileSystem} fs
 */
const fs: FileSystem = {
  /**
   * Get the contents of a file.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  readFileSync(): never {
    throw new Error('[readFileSync] not implemented')
  },

  /**
   * Read the contents of a directory.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  readdirSync(): never {
    throw new Error('[readdirSync] not implemented')
  },

  /**
   * Get the resolved pathname of a directory or file.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  realpathSync(): never {
    throw new Error('[realpathSync] not implemented')
  },

  /**
   * Get information about a directory or file.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  statSync(): never {
    throw new Error('[statSync] not implemented')
  }
}

export default fs
