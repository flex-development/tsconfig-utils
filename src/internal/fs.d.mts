declare module '#internal/fs' {
  import type { FileSystem } from '@flex-development/tsconfig-utils'

  /**
   * File system API.
   *
   * @const {FileSystem} fs
   */
  const fs: FileSystem

  export default fs
}
