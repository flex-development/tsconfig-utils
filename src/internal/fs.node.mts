/**
 * @file Internal - fs/node
 * @module mkbuild/internal/fs/node
 */

import type { FileSystem } from '@flex-development/tsconfig-utils'
import { readdirSync, readFileSync, realpathSync, statSync } from 'node:fs'

/**
 * File system API.
 *
 * @const {FileSystem} fs
 */
const fs: FileSystem = {
  readFile: readFileSync,
  readdir: readdirSync,
  realpath: realpathSync,
  stat: statSync
}

export default fs
