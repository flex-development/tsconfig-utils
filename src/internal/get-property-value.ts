/**
 * @file Internal - getPropertyValue
 * @module tsconfig-utils/internal/getPropertyValue
 */

import type { Nilable, ObjectUnknown } from '@flex-development/tutils'

/**
 * Extracts a property value from the given `object`.
 *
 * Returns `undefined` if the given `property` is not defined.
 *
 * **Note**: If `object` is `null` or `undefined`, it will be converted to an
 * empty object first.
 *
 * @template T - Object type
 * @template K - Object property name
 *
 * @param {Nilable<T>} object - Object to evaluate, `null`, or `undefined`
 * @param {string} property - Object property name
 * @return {T[K] | undefined} Property value or `undefined`
 */
function getPropertyValue<T extends ObjectUnknown, K extends string = string>(
  object: Nilable<T>,
  property: K
): T[K] | undefined {
  return Object.getOwnPropertyDescriptor(object ?? {}, property)?.value as T[K]
}

export default getPropertyValue
