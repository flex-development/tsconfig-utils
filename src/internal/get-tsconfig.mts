/**
 * @file Internal - getTsconfig
 * @module tsconfig-utils/internal/getTsconfig
 */

import isTsconfigHost from '#lib/is-tsconfig-host'
import type { Tsconfig, TsconfigHost } from '@flex-development/tsconfig-utils'
import { shake } from '@flex-development/tutils'

/**
 * Get a {@linkcode Tsconfig} from `target`.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Tsconfig | TsconfigHost | null | undefined} target
 *  Tsconfig or tsconfig host
 * @return {Tsconfig}
 *  TypeScript configuration object
 */
function getTsconfig(
  this: void,
  target: Tsconfig | TsconfigHost | null | undefined
): Tsconfig {
  return shake(isTsconfigHost(target) ? target.tsconfig : target ?? {})
}

export default getTsconfig
