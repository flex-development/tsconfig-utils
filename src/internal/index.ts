/**
 * @file Entry Point - Internal
 * @module tsconfig-utils/internal
 */

export { default as isDirectory } from './is-directory'
export { default as isFile } from './is-file'
export { default as readFile } from './read-file'
export { default as validateFunction } from './validate-function'
export { default as validateString } from './validate-string'
export { default as validateURLString } from './validate-url-string'
