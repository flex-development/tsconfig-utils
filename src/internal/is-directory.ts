/**
 * @file Internal - isDirectory
 * @module tsconfig-utils/internal/isDirectory
 */

import type mlly from '@flex-development/mlly'
import fs from 'node:fs'
import validateURLString from './validate-url-string'

/**
 * Checks if a directory exists at the given module `id`.
 *
 * @param {mlly.ModuleId} id - Module id to evaluate
 * @return {boolean} `true` if directory exists at `id`
 */
const isDirectory = (id: mlly.ModuleId): boolean => {
  try {
    validateURLString(id, 'id')
    return fs.statSync(id, { throwIfNoEntry: false })?.isDirectory() ?? false
  } catch {
    return false
  }
}

export default isDirectory
