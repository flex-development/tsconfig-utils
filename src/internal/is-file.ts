/**
 * @file Internal - isFile
 * @module tsconfig-utils/internal/isFile
 */

import type mlly from '@flex-development/mlly'
import fs from 'node:fs'
import validateURLString from './validate-url-string'

/**
 * Checks if a file exists at the given module `id`.
 *
 * @param {mlly.ModuleId} id - Module id to evaluate
 * @return {boolean} `true` if file exists at `id`
 */
const isFile = (id: mlly.ModuleId): boolean => {
  try {
    validateURLString(id, 'id')
    return fs.statSync(id, { throwIfNoEntry: false })?.isFile() ?? false
  } catch {
    return false
  }
}

export default isFile
