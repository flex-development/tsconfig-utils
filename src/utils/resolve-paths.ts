/**
 * @file Utilities - resolvePaths
 * @module tsconfig-utils/utils/resolvePaths
 */

import type { ResolvePathsOptions } from '#src/interfaces'
import * as internal from '#src/internal'
import * as mlly from '@flex-development/mlly'
import type { CompilerOptions } from '@flex-development/tsconfig-types'
import loadCompilerOptions from './load-compiler-options'

/**
 * Resolves path aliases in `export`, `import`, and `require` statements in the
 * given piece of source `code`.
 *
 * @async
 *
 * @param {string} code - Code to evaluate
 * @param {ResolvePathsOptions} options - Path alias resolution options
 * @return {Promise<string>} `code` with path aliases resolved and/or unmodified
 */
const resolvePaths = async (
  code: string,
  options: ResolvePathsOptions
): Promise<string> => {
  const {
    absolute = false,
    baseUrl,
    condition,
    conditions,
    ext,
    extensions,
    file = internal.isFile,
    parent,
    preserveSymlinks,
    read = internal.readFile,
    tsconfig = mlly.toURL('tsconfig.json')
  } = options

  /**
   * User compiler options.
   *
   * @const {CompilerOptions} compilerOptions
   */
  const compilerOptions: CompilerOptions = loadCompilerOptions(tsconfig, {
    file,
    read
  })

  return mlly.resolveAliases(code, {
    absolute,
    aliases: compilerOptions.paths,
    condition,
    conditions,
    cwd: baseUrl ?? compilerOptions.baseUrl,
    ext,
    extensions,
    parent,
    preserveSymlinks
  })
}

export default resolvePaths
