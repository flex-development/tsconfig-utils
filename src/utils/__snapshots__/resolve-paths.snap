// Vitest Snapshot v1

exports[`unit:utils/resolvePaths > should return code with path aliases resolved 1`] = `
"import type { ResolvePathsOptions } from '../interfaces'
import * as internal from '../internal'
import * as mlly from '@flex-development/mlly'
import type { CompilerOptions } from '@flex-development/tsconfig-types'
import loadCompilerOptions from './load-compiler-options'"
`;
